java/integTest/src/test/java/glide/PubSubTests.java

```diff
@@ -0,0 +1,1013 @@
+                                    .build())
+                    .get();
+        }
+    }
+
+    private <M extends ChannelMode> BaseClient createClientWithSubscriptions(
+            boolean standalone, Map<M, Set<String>> subscriptions) {
+        return createClientWithSubscriptions(
+                standalone, subscriptions, Optional.empty(), Optional.empty());
+    }
+
+    @SneakyThrows
+    private BaseClient createClient(boolean standalone) {
+        if (standalone) {
+            return RedisClient.CreateClient(commonClientConfig().build()).get();
+        }
+        return RedisClusterClient.CreateClient(commonClusterClientConfig().build()).get();
+    }
+
+    /**
+     * PubsubMessage queue used in callback to analyze received PubsubMessages. Number is a client ID.
+     */
+    private final ConcurrentLinkedDeque<Pair<Integer, PubSubMessage>> PubsubMessageQueue =
+            new ConcurrentLinkedDeque<>();
+
+    /** Clients used in a test. */
+    private final List<BaseClient> clients = new ArrayList<>();
+
+    @BeforeEach
+    @SneakyThrows
+    public void cleanup() {
+        for (var client : clients) {
+            if (client instanceof RedisClusterClient) {
+                ((RedisClusterClient) client).customCommand(new String[] {"unsubscribe"}, ALL_NODES).get();
+                ((RedisClusterClient) client).customCommand(new String[] {"punsubscribe"}, ALL_NODES).get();
+                ((RedisClusterClient) client).customCommand(new String[] {"sunsubscribe"}, ALL_NODES).get();
+            } else {
+                ((RedisClient) client).customCommand(new String[] {"unsubscribe"}).get();
+                ((RedisClient) client).customCommand(new String[] {"punsubscribe"}).get();
+            }
+            client.close();
+        }
+        clients.clear();
+        PubsubMessageQueue.clear();
+    }
+
+    private void verifyReceivedPubsubMessages(
+            Set<Pair<Integer, PubSubMessage>> PubsubMessages, BaseClient listener, boolean callback) {
+        if (callback) {
+            assertEquals(PubsubMessages, new HashSet<>(PubsubMessageQueue));
+        } else {
+            var received = new HashSet<PubSubMessage>(PubsubMessages.size());
+            PubSubMessage PubsubMessage;
+            while ((PubsubMessage = listener.tryGetPubSubMessage()) != null) {
+                received.add(PubsubMessage);
+            }
+            assertEquals(
+                    PubsubMessages.stream().map(Pair::getValue).collect(Collectors.toSet()), received);
+        }
+    }
+
+    public static Stream<Arguments> getTwoBoolPermutations() {
+        return Stream.of(
+                Arguments.of(true, true),
+                Arguments.of(true, false),
+                Arguments.of(false, true),
+                Arguments.of(false, false));
+    }
+
+    // TODO add following tests from https://github.com/aws/glide-for-redis/pull/1643
+    //  test_pubsub_exact_happy_path_coexistence
```

Comment:
still todo?

Reply:
Those are blocked by `getPubSubMessage`, which is not implemented